Feature/exchange flow (#52)
hermeznetwork/docs: docs/users/exchanges.md

# Exchanges

This example shows a possible flow of how an exchange would use Hermez.

## Pre-requisites
-  Exchange already has some Hermez accounts for each trading token (`HEZEx-ETH`, `HEZEx-DAI`,...). 

> -  User has an Hermez account (`HEZUser-ETH`)

Reviewer: ```suggestion
-  User has a Hermez account (`HEZUser-ETH`)
```

![](exchanges/Exchange-init.jpg)

Existing accounts are regular Hermez accounts, consisting of an L2 account linked to an Ethereum account where funds can be withdrawn.

## User Transfers Tokens to Exchange
A User wants to transfer 10 ETH from his Hermez account to the exchange for the first time. To do so, the user requests to transfer some funds via some sort of front end provided by the exchange.
After the request is done, the exchange provides the address of an internal Hermez account where the user can deposit his tokens. This account doesn't have an Ethereum counterpart account, 

> and thus the creation is really cheap. Once the user has received the L2 account address, he can perform the transfer normally.

Reviewer: ```suggestion
and thus the creation is very inexpensive. Once the user has received the L2 account address, he can perform the transfer normally.
```

In the meantime, the exchange is monitoring the status of this account, and once the user transfer is completed, the exchange can transfer the funds to its main account. This process is depicted in the diagram below.

The creation of this user account by the exchange is only done once. 

![](exchanges/Exchange-transfer.jpg)

### Flow 

> 1. User requests to do a transfer to the exchange from his Hermez account using some front-end.

Reviewer: Please discard if this isn't the correct terminology for this instance:
"1. User requests to do a transfer to the exchange from his Hermez account using a front-end framework."

> 2. Exchange creates an L2 (internal) account on behalf of the user (`L2ExUser-ETH`) and provides the address via a front-end. This account is controlled by the exchange. 

Reviewer: ```suggestion
2. Exchange creates an L2 (internal) account on behalf of the user (`L2ExUser-ETH`) and provides the address via front-end framework. This account is controlled by the exchange. 
```

```js
  const infoAccountUser = (await hermez.CoordinatorAPI.getAccounts(hermezUserWallet.hermezEthereumAddress, [tokenERC20.id]))
    .accounts[0];

  const state = await hermez.CoordinatorAPI.getState();
  const usdTokenExchangeRate = tokenERC20.USD;
  const fee = usdTokenExchangeRate ? state.recommendedFee.createAccountInternal / usdTokenExchangeRate : 0;

  // user creates transaction to deposit 10 ether into exchange account
  // deposit 10 ether
  const userDepositToExchange = hermez.Utils.getTokenAmountBigInt("10.0", 18);
  const compressedUserDepositToExchange = hermez.HermezCompressedAmount.compressAmount(userDepositToExchange);
  // the following transaction would:
  // - create an account for the exchange in hermez network
  // - transfer to exchange account 0.1 eth
  const transferToExchange = {
    from: infoAccountUser.accountIndex,
    to: hermezExchangeUserWallet.publicKeyBase64,
    amount: compressedUserDepositToExchange,
    fee : fee
  };
  console.log("transferToExchange: ", transferToExchange, fee);
  // send tx to hermez network
  await hermez.Tx.generateAndSendL2Tx(transferToExchange, hermezUserWallet, tokenERC20);
```

> 4. Exchange monitors balance of the account `L2ExUser-ETH`, and once transfer has completed, exchange performs transfer from `L2ExUser-ETH` to `L2Ex-ETH` for 10 ETH.

Reviewer: ```suggestion
4. Exchange monitors balance of the account `L2ExUser-ETH`, and once transfer is complete, exchange performs transfer from `L2ExUser-ETH` to `L2Ex-ETH` for 10 ETH.
```

    tokenERC20,
    hermezUserWallet.publicKeyCompressedHex,
    { type: "WALLET", privateKey: userPrivKey }
  );

  // WAIT until accounts are created
  const pollingAccountCreate = true;
  while (pollingAccountCreate){
    const accountExchangeInfo = await hermez.CoordinatorAPI.getAccounts(hermezExchangeEthereumAddress, [tokenERC20.id]);
    if (accountExchangeInfo.accounts.length === 0){
      console.log("Waiting for deposits to be forged...");
      await sleep(10000);
    } else {
      break;
    }

  const infoAccountExchange = (await hermez.CoordinatorAPI.getAccounts(hermezExchangeWallet.hermezEthereumAddress, [tokenERC20.id]))
    .accounts[0];
  
  // EXCHANGE ACTION
  // create new bjj private key to receive user transactions
  const pvtBjjKey = Buffer.allocUnsafe(32).fill("1");

  // create rollup internal account from bjj private key
  const resExchangeWallet = await hermez.HermezWallet.createWalletFromBjjPvtKey(pvtBjjKey);
  const hermezExchangeUserWallet = resExchangeWallet.hermezWallet;

  // share public bjj key with the user
  console.log(`Transfer funds to this hermez address:\n   ${hermezExchangeUserWallet.publicKeyBase64}\n\n`);

  // USER ACTION

>   // - the following code could be done through the web wallet provided by hermez network

Reviewer: ```suggestion
  // - the following code could be done through the web wallet provided by Hermez Network
```

>   // - it is assumed that the user has already ether in hermez network

Reviewer: ```suggestion
  // - it is assumed that the user has already Ether in Hermez Network
```

# Exchanges

This example shows a possible flow of how an exchange would use Hermez.

## Pre-requisites

> -  Exchange already has some Hermez accounts for each trading token (`HEZEx-ETH`, `HEZEx-DAI`,...). 

Reviewer 2: it could be useful to add link here to "how to create accounts" documentation

> -  User has a Hermez account (`HEZUser-ETH`)

Reviewer 2: ```suggestion
-  User has an Hermez account (`HEZUser-ETH`)
```

![](exchanges/Exchange-init.jpg)

> Both exchange and user p pre-existing accounts are regular Hermez accounts, consisting of an L2 account linked to an Ethereum account where funds can be withdrawn.

Reviewer 2: ```suggestion
Both exchange and user pre-existing accounts are regular Hermez accounts, consisting of an L2 account linked to an Ethereum account where funds can be withdrawn.
```

> The following is an example of code used to initialize hermezjs` with `Hermez testnet` deployment and create both user and exchange accounts. You need to supply your

Reviewer 2: ```suggestion
The following is an example of code used to initialize hermezjs with `Hermez testnet` deployment and create both user and exchange accounts. You need to supply your
```

> own WEB3_URL (Ethereum Node URL) and two Rinkeby Ethereum Provate Keys.

Reviewer 2: ```suggestion
own WEB3_URL (Ethereum Node URL) and two Rinkeby Ethereum Private Keys.
```

const EXAMPLES_HERMEZ_WDELAYER_ADDRESS = "0x6ea0abf3ef52d24427043cad3ec26aa4f2c8e8fd";

// Provide your own values
const EXAMPLES_WEB3_URL = ""
const EXAMPLES_PRIVATE_KEY1 = ""
const EXAMPLES_PRIVATE_KEY2 = ""

async function sleep (timeout) {
  await new Promise(resolve => setTimeout(resolve, timeout));
}

function configureEnvironment () {
  // Initializes Tx Pool
  hermez.TxPool.initializeTransactionPool()
  // load ethereum network provider
  hermez.Providers.setProvider(EXAMPLES_WEB3_URL)

  // set environment
  hermez.Environment.setEnvironment({
    baseApiUrl: EXAMPLES_HERMEZ_API_URL,
    contractAddresses: {
      [hermez.Constants.ContractNames.Hermez]: EXAMPLES_HERMEZ_ROLLUP_ADDRESS,
      [hermez.Constants.ContractNames.WithdrawalDelayer]: EXAMPLES_HERMEZ_WDELAYER_ADDRESS
    }
  })
}

async function main(){
  // INITIALIZATION
  // initialize hermezjs and prepare two Hermez accounts with some ETH (user and exchange accounts)
  configureEnvironment()

  const exchangePrivKey = EXAMPLES_PRIVATE_KEY1;
  const userPrivKey = EXAMPLES_PRIVATE_KEY2;

  // load token to deposit information
  const tokenIndex = 0;
  const token = await hermez.CoordinatorAPI.getTokens();

>   const tokenERC20 = token.tokens[tokenIndex];

Reviewer 2: Token index 0 is `ether`. I may be useful to change this name to `tokenEth` or just a better appropriate name

  // load first account
  const wallet = await hermez.HermezWallet.createWalletFromEtherAccount(EXAMPLES_WEB3_URL, { type: "WALLET", privateKey: exchangePrivKey });
  const hermezExchangeWallet = wallet.hermezWallet;
  const hermezExchangeEthereumAddress = wallet.hermezEthereumAddress;

  // load second account
  const wallet2 = await hermez.HermezWallet.createWalletFromEtherAccount(EXAMPLES_WEB3_URL, { type: "WALLET", privateKey: userPrivKey });
  const hermezUserWallet = wallet2.hermezWallet;
  const hermezUserEthereumAddress = wallet2.hermezEthereumAddress;

>   console.log("W2", wallet2)

Reviewer 2: I would remove this `console.log`

  // set amount to deposit
  const amountDeposit = hermez.Utils.getTokenAmountBigInt("0.1", 18);
  const compressedDepositAmount = hermez.HermezCompressedAmount.compressAmount(amountDeposit);

  // perform deposit hermezExchangeAccount
  await hermez.Tx.deposit(
    compressedDepositAmount,
    hermezExchangeEthereumAddress,
    tokenERC20,
    hermezExchangeWallet.publicKeyCompressedHex,
    { type: "WALLET", privateKey: exchangePrivKey }
  );

  // perform deposit hermezUserAccount
  await hermez.Tx.deposit(
    compressedDepositAmount,
    hermezUserEthereumAddress,
    tokenERC20,
    hermezUserWallet.publicKeyCompressedHex,
    { type: "WALLET", privateKey: userPrivKey }
  );

  // WAIT until accounts are created
  const pollingAccountCreate = true;
  while (pollingAccountCreate){
    const accountExchangeInfo = await hermez.CoordinatorAPI.getAccounts(hermezExchangeEthereumAddress, [tokenERC20.id]);
    if (accountExchangeInfo.accounts.length === 0){
      console.log("Waiting for deposits to be forged...");
      await sleep(10000);
    } else {
      console.log("Accounts created", accountExchangeInfo)
      break;
    }
  }

>   const infoAccountExchange = (await hermez.CoordinatorAPI.getAccounts(hermezExchangeWallet.hermezEthereumAddress, [tokenERC20.id]))

Reviewer 2: It could be useful to add a `console.log("infoAccountExchange: ", infoAccountExchange)` here.
Otherwise, I would remove this line since it is not used

    .accounts[0];

}

main();
```

> ## User Transfers Tokens to Exchange

Reviewer 2: Title `User Transfers Tokens to Exchange` is not shown properly when rendering hackmd in web. Seems that title is too large for content search panel on the right side. I would suggest to change the name for a shorter one such as `Deposit to exchange`

A User wants to transfer 10 ETH from his Hermez account to the exchange for the first time. To do so, the user requests to transfer some funds via some sort of front end provided by the exchange.
After the request is done, the exchange provides the address of an internal Hermez account where the user can deposit his tokens. This account doesn't have an Ethereum counterpart account, 
and thus the creation is very inexpensive. Once the user has received the L2 account address, he can perform the transfer normally.
In the meantime, the exchange is monitoring the status of this account, and once the user transfer is completed, the exchange can transfer the funds to its main account. This process is depicted in the diagram below.

> The creation of this user account by the exchange is only done once. 

Reviewer 2: It is done once per user and token id. Or it could not be done once.
Exchange could provide different addresses for the same user-tokenId. I think it would depend exclusively on exchange decision.
I would remove this sentence

    from: infoAccountUser.accountIndex,
    to: hermezExchangeUserWallet.publicKeyBase64,
    amount: compressedUserDepositToExchange,
    fee : fee
  };
  console.log("transferToExchange: ", transferToExchange, fee);
  // send tx to hermez network
  await hermez.Tx.generateAndSendL2Tx(transferToExchange, hermezUserWallet, tokenERC20);
```
4. Exchange monitors balance of the account `L2ExUser-ETH`, and once transfer is complete, exchange performs transfer from `L2ExUser-ETH` to `L2Ex-ETH` for 10 ETH.
```js
  const pollingExchangeAddr = true;
  while (pollingExchangeAddr){
    const accountExchangeInfo = await hermez.CoordinatorAPI.getAccounts(hermezExchangeUserWallet.publicKeyBase64, [tokenERC20.id]);
    if (accountExchangeInfo.accounts.length === 0){
      console.log("Waiting for user deposit to be forged...");
      await sleep(10000);
    } else {
      console.log("<=== Received deposit from user ===>");
      console.log(`accountExchangeInfo:\n ${accountExchangeInfo.accounts[0]}`);
      break;
    }
  }

  const infoAccountExchangeUser = (await hermez.CoordinatorAPI.getAccounts(hermezExchangeUserWallet.publicKeyBase64, [tokenERC20.id]))
   .accounts[0];

>   // Transfer funds to main exchange account

Reviewer 2: The following code is rendered well in web.
Delete one line and it will be rendered properly

  // generate L2 transaction
  const l2TxTransfer = {
    from: infoAccountExchangeUser.accountIndex,
    to:  infoAccountExchange.accountIndex,
    amount: compressedUserDepositToExchange,
    fee: fee
  };

  const transferResponse = await hermez.Tx.generateAndSendL2Tx(l2TxTransfer, hermezExchangeUserWallet, tokenERC20).catch(console.log);
  console.log("transferResponse: ", transferResponse);
```

### Full Example
```js
const hermez = require("@hermeznetwork/hermezjs");

const EXAMPLES_HERMEZ_API_URL = "https://api.testnet.hermez.io/";
const EXAMPLES_HERMEZ_ROLLUP_ADDRESS = "0x14a3b6f3328766c7421034e14472f5c14c5ba090";
const EXAMPLES_HERMEZ_WDELAYER_ADDRESS = "0x6ea0abf3ef52d24427043cad3ec26aa4f2c8e8fd";

// Provide your own values

> const EXAMPLES_WEB3_URL = "http://----";

Reviewer 2: In section `pre-requisites`, those variables are empty.
I would either change `pre-requisites` section or this one in order to match them
